stac_fastapi/pgstac/stac_fastapi/pgstac/utils.py

```diff
@@ -0,0 +1,120 @@
+"""stac-fastapi utility methods."""
+from copy import deepcopy
+from typing import Any, Dict, Optional, Set, Union
+
+from stac_fastapi.pgstac.types.base_item_cache import BaseItemCache
+from stac_fastapi.types.stac import Item
+
+
+async def hydrate(
+    item: Union[Item, Dict[str, Any]], base_item_cache: BaseItemCache
+) -> Item:
+    """Hydrate item in-place with base_item properties.
+
+    This will not perform a deep copy; values of the original item will be referenced
+    in the return item.
+    """
+    item = dict(item)
+
+    # Merge will mutate i, but create deep copies of values in the base item
+    # This will prevent the base item values from being mutated, e.g. by
+    # filtering out fields in `filter_fields`.
+    def merge(b: Dict[str, Any], i: Dict[str, Any]):
+        for key in b:
+            if key in i:
+                if isinstance(b[key], dict) and isinstance(i.get(key), dict):
+                    # Recurse on dicts to merge values
+                    merge(b[key], i[key])
+                elif b[key] == i.get(key):
+                    # Matching key/value is a no-op
+                    pass
+                elif isinstance(b[key], list) and isinstance(i.get(key), list):
+                    # Merge unequal lists
+                    i[key].extend(deepcopy(b[key]))
```

Comment:
What's the reason for the `deepcopy` rather than something like

```python
i[key].extend(b[key])
```

I suppose then we're going to later mutate `i[key]`, and we don't want to mutate the base `b[key]`?

Can you confirm that we don't have cases where we repeatedly copy the same data? i.e. we do a `deepcopy` on `b[key]`, but then we later do a `deepcopy` on `b[key][0]`, `b[key][1]`, ... as we call `merge` recursively?

Reply:
That's correct, `i[key]` may be mutated later and `b[key]` may have been applied to many instances of `i`.

I can confirm that at the point of the `deepcopy` (in the new implementation in pypgstac), the code will never recurse deeper into any paths beneath `b[key]` so we shouldn't be copying the same data more than once.